clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -1231,66 +1248,71 @@ protected void produceToStoreBufferServiceOrKafkaInBatch(
     List<List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> batches = new ArrayList<>();
     List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> ongoingBatch = new ArrayList<>(batchSize);
     Iterator<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> iter = records.iterator();
-    while (iter.hasNext()) {
-      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> record = iter.next();
-      if (partitionConsumptionState != null) {
+
+    /*
+     * Acquire the read lock to ensure that the result of shouldProcessRecord() holds for the entire duration for
+     * which this message is processed. This lock protects against any state transitions where the SIT needs to
+     * acquire the write lock to modify the leader-follower state, since it would need to wait for this to finish.
+     */
+    final ReadWriteLock leaderFollowerStateLock = partitionConsumptionState.getLeaderFollowerStateLock();
+    try (AutoCloseableLock ignore = AutoCloseableSingleLock.of(leaderFollowerStateLock.readLock())) {
```

Comment:
Here you acquire lock for entire loop but inside `produceToStoreBufferServiceOrKafka` lock is acquired for every message processing, maybe there is a reason for this?

Reply:
Because the condition of `shouldProcessMessage()` needs to hold for the entire duration for which the message is processed by the consumer.

In the batch version (`produceToStoreBufferServiceOrKafkaInBatch()`), the condition is checked to determine the batches of selected messages, and then processing begins with `ingestionBatchProcessor.process()`. Thus, it is easiest and safest to protect the entire section with the read lock.